apps/src/music/blockly/toolbox/index.ts

```diff
@@ -110,8 +111,176 @@ export function getToolbox(
+ */
+export const toolboxModeCategory = {
+  kind: 'category',
+  name: 'Categories',
+  cssconfig: baseCategoryCssConfig,
+  contents: [
+    {kind: 'block', type: BlockTypes.CATEGORY},
+    {kind: 'block', type: BlockTypes.CUSTOM_CATEGORY},
+  ],
+};
+
+/**
+ * Given a toolbox definition and a workspace, iterate through any
+ * categories, add all blocks to the workspace. Used for levelbuilder's
+ * toolbox mode.
+ */
+export function addToolboxBlocksToWorkspace(
+  contents: GoogleBlockly.utils.toolbox.ToolboxItemInfo[],
+  workspace: GoogleBlockly.WorkspaceSvg,
+  clearWorkspace: boolean = true
+) {
+  if (clearWorkspace) {
+    workspace.clear();
+  }
+  contents.forEach(toolboxItem => {
+    if (toolboxItem.kind === 'block') {
+      // Add toolbox blocks directly to the workspace.
+      Blockly.serialization.blocks.append(
+        toolboxItem as GoogleBlockly.serialization.blocks.State,
+        workspace
+      );
+    } else if (toolboxItem.kind === 'category' && 'custom' in toolboxItem) {
+      // For dynamic categories, create a custom category block..
+      const dynamicCategoryName = (
+        toolboxItem as GoogleBlockly.utils.toolbox.DynamicCategoryInfo
+      ).id;
+      Blockly.serialization.blocks.append(
+        {
+          type: BlockTypes.CUSTOM_CATEGORY,
+          fields: {
+            CUSTOM: dynamicCategoryName,
+          },
+        },
+        workspace
+      );
+    } else if (toolboxItem.kind === 'category') {
+      const categoryInfo =
+        toolboxItem as GoogleBlockly.utils.toolbox.StaticCategoryInfo;
+      // For a localized category, like "Sounds", get the category type, like "Play".
+      const categoryName = categoryInfo.id;
+      // 'DEFAULT' categories are intentionally skipped.
+      if (categoryName && categoryName in Category) {
+        // Create a category block
+        Blockly.serialization.blocks.append(
+          {
+            type: BlockTypes.CATEGORY,
+            fields: {
+              CATEGORY: categoryName,
+            },
+          },
+          workspace
+        );
+      } else {
+        // We shouldn't hit this point, but theoretically could if a toolbox
+        // contained unsupported items like buttons or labels.
+        console.warn('Unsupported category found:', toolboxItem);
+      }
+      // Recursively process the contents of the static category
+      addToolboxBlocksToWorkspace(
+        (toolboxItem as GoogleBlockly.utils.toolbox.StaticCategoryInfo)
+          .contents,
+        workspace,
+        false
+      );
+    } else {
+      console.warn('Unsupported toolbox item found:', toolboxItem);
+    }
+  });
+}
+
+/**
+ * Creates a new static category. Used to convert workspace
+ * blocks into a toolbox definition in levelbuilder's toolbox mode.
+ * @returns JSON representation of a new static category.
+ */
+export function getNewStaticCategory(
+  name: string = 'DEFAULT'
```

Comment:
After seeing your live demo, I'm wondering if this should be something a little more specific like `UNCATEGORIZED` or `WARNING: NO CATEGORY` or something, so it's a little more obvious what will happen to these blocks?

Also minor: create a constant for this since we check against it elsewhere?

Reply:
I went with 'DEFAULT' here to align to other labs. It _is_ technically a category, so I think this might be the clearest. Good call on creating a constant.